docs/architecture/adr-056-proving-amnesia-attack.md

```diff
@@ -0,0 +1,106 @@
+# ADR 056: Proving flip flopping attacks
+
+## Changelog
+
+- 02.04.20: Initial Draft
+- 06.04.20: Second Draft
+
+## Context
+
+Whilst most created evidence of malicious behaviour is self evident such that any individual can verify them independently there are types of evidence, known as global evidence, that require further collaboration from the network in order to accumulate enough information to create evidence that is individually verifiable and can therefore be processed through consensus. Fork Accountability as a whole constitutes of detection, proving and punishing. This ADR addresses how to prove infringement from global evidence that is sent to a full node.
+
+The currently only known form of global evidence stems from [flip flopping](https://github.com/tendermint/spec/blob/master/spec/consensus/light-client/accountability.md#flip-flopping) attacks. The schematic below explains one scenario where an amnesia attack, a form of flip flopping, can occur such that C1 and C2 commit different blocks.
+
+![](../imgs/tm-amnesia-attack.png)
+
+This creates a fork on the main chain.  Back to the past, another form of flip flopping, creates a light fork (capable of fooling those not involved in consensus), in a similar way, with F taking the precommits from C1 and forging a commit from them.
+
+## Decision
+
+As the distinction between these two attacks (amnesia and back to the past) can only be distinguished by confirming with all validators (to see if it is a full fork or a light fork), for the purpose of simplicity, these attacks will be treated as the same.
+
+Currently, the evidence reactor is used to simply broadcast and store evidence. Instead of perhaps creating a new reactor for the specific task of verifying these attacks, the current evidence reactor will be extended.
+
+The process begins with a light client receiving conflicting headers (in the future this could also be a full node during fast sync), which it sends to a full node to analyse. As part of [evidence handling](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-047-handling-evidence-from-light-client.md), this could be deduced into potential amnesia evidence
+
+```
+type PotentialAmnesiaEvidence struct {
```

Comment:
If we indeed use "flipflop" instead of amnesia, let's call this `PotentialFlipFlopEvidence` or something instead

Reply:
@melekes what do you think?

Reply:
**I think we should stick with PotentialAmnesiaEvidence. It's consistent with AmnesiaEvidence, which is created as a result.** It's true that PotentialAmnesiaEvidence can contain Back to the past evidence, making it FlipFloppingEvidence in reality, but we only punish Amnesia evidence.

Reply:
Okay :+1: I still fail to see how we distinguish between Back to the past and amnesia, both of these break the locking mechanism and that's what we detect when we see different commits at different rounds

Reply:
I'd keep `PotentialAmnesiaEvidence -> AmnesiaEvidence` or introduce `FlipFloppingEvidence -> AmnesiaEvidence` (note no Potential prefix since flip flopping is there, you just don't know which is it). Both are fine, but I slightly prefer (1).